flink-runtime/src/test/java/org/apache/flink/runtime/jobmaster/JobMasterTest.java

```diff
@@ -0,0 +1,164 @@
+
+		final TestingSerialRpcService rpc = new TestingSerialRpcService();
+		rpc.registerGateway(taskManagerAddress, taskExecutorGateway);
+
+		final ScheduledExecutorService executorService = Executors.newScheduledThreadPool(
+				Hardware.getNumberCPUCores(), ExecutorThreadFactory.INSTANCE);
+
+		final long heartbeatInterval = 10L;
+		final long heartbeatTimeout = 1000L;
+		final HeartbeatManagerSenderImpl<Object, Object> jmHeartbeatManager = new HeartbeatManagerSenderImpl<>(
+				heartbeatInterval,
+				heartbeatTimeout,
+				jmResourceId,
+				executorService,
+				rpc.getScheduledExecutor(),
+				log);
+
+		final JobVertex jobVertex = new JobVertex("NoOpInvokable");
+		jobVertex.setInvokableClass(NoOpInvokable.class);
+		final JobGraph jobGraph = new JobGraph("test", jobVertex);
+
+		final BlobLibraryCacheManager libraryCacheManager = new BlobLibraryCacheManager(mock(BlobService.class), 1000000000L);
+		libraryCacheManager.registerJob(jobGraph.getJobID(), jobGraph.getUserJarBlobKeys(), jobGraph.getClasspaths());
+
+		final MetricRegistry registry = new MetricRegistry(MetricRegistryConfiguration.defaultMetricRegistryConfiguration());
+		final JobManagerMetricGroup jmMetricGroup = new JobManagerMetricGroup(registry, "host");
+
+		try {
+			final JobMaster jobMaster = new JobMaster(
+					jobGraph,
+					new Configuration(),
+					rpc,
+					haServices,
+					executorService,
+					libraryCacheManager,
+					mock(RestartStrategyFactory.class),
+					Time.of(10, TimeUnit.SECONDS),
+					jmMetricGroup,
+					jmResourceId,
+					jmHeartbeatManager,
+					mock(OnCompletionActions.class),
+					testingFatalErrorHandler,
+					libraryCacheManager.getClassLoader(jobGraph.getJobID()));
+
+			// also start the heartbeat manager in job manager
+			jobMaster.start(jmLeaderId);
+
+			// register task manager will trigger monitoring heartbeat target, schedule heartbeat request in interval time
+			jobMaster.registerTaskManager(taskManagerAddress, taskManagerLocation, jmLeaderId);
+
+			final ConcurrentHashMap<ResourceID, Object> heartbeatTargets = Whitebox.getInternalState(jmHeartbeatManager, "heartbeatTargets");
+			final Map<ResourceID, Tuple2<TaskManagerLocation, TaskExecutorGateway>> registeredTMsInJM = Whitebox.getInternalState(jobMaster, "registeredTaskManagers");
+			final SlotPool slotPool = Whitebox.getInternalState(jobMaster, "slotPool");
+			final HashSet<ResourceID> registeredTMsInSlotPool = Whitebox.getInternalState(slotPool, "registeredTaskManagers");
+
+			// before heartbeat timeout
+			assertTrue(heartbeatTargets.containsKey(tmResourceId));
+			assertTrue(registeredTMsInJM.containsKey(tmResourceId));
+			assertTrue(registeredTMsInSlotPool.contains(tmResourceId));
+
+			// trigger heartbeat timeout in job manager side, because the task manager will not response the heartbeat
+			Thread.sleep(heartbeatTimeout);
+
+			// after heartbeat timeout
+			verify(taskExecutorGateway, atLeast(1)).heartbeatFromJobManager(eq(jmResourceId));
```

Comment:
In order to wait for the timeout being triggered you could extend the `HeartbeatManagerImpl` and introduce a wait mechanism in `unmonitorTarget` method. This should be called by the `JobMaster` to unregister the timed out `TaskExecutor`.